### PR TITLE
New merge lists plugin from a given key

### LINKED ISSUE #249: New merge lists plugin from a given key
##### SUMMARY

Given two lists like:

```yaml
list1:
  - name: myname01
    param01: myparam01
  - name: myname02
    param01: myparam02

list2:
  - name: myname01
    param01: myparam03
  - name: myname02
    param02: myparam04
  - name: myname03
    param03: myparam03
```

When doing:

```yaml
"{{ query('mergelist', 'name', list1, list2) }}"
```

The results will be:

```yaml
  - name: myname01
    param01: myparam03
  - name: myname02
    param01: myparam02
    param02: myparam04
  - name: myname03
    param03: myparam03
```

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

`mergelist`
##### ADDITIONAL INFORMATION

I have already developed the solution. I really think it could be really useful in the Ansible Collections General. Please confirm.

```yaml
  - name: Basic usage
    debug:
      msg: "{{ query('mergelist', 'name', list1, list2) }}"
```

### PR DESCRIPTION
**Pull request by imjoseangel**
##### SUMMARY

Fixes #249 
##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

mergelist

##### ADDITIONAL INFORMATION

When creating complex and long structures there are ways to merge or replace one to one dictionaries starting from their head. There is no a simple way to merge small fragments of different dictionaries from a given master key.

For instance: A keyvault ACL in Azure can be unreadable. To keep a consistency and merge new configurations and don't break the existing ones, something like this can be performed:

```yaml
q('mergelist',
       _keyvaultinfo.keyvaults[0].access_policies | default([]),
        keyvault_objectid.keyvaultsp | default([]) ) |
          json_query('[*].{
            tenant_id: tenant_id, 
            object_id: object_id,
            certificates: permissions.certificates,
            keys: permissions.keys,
            secrets: permissions.secrets
            }', 
           key='object_id')
```
Something like this without this plugin is not easily achievable.

Given two lists like:

```yaml
list1:
  - name: myname01
    param01: myparam01
  - name: myname02
    param01: myparam02

list2:
  - name: myname01
    param01: myparam03
  - name: myname02
    param02: myparam04
  - name: myname03
    param03: myparam03
```

When doing:

```yaml
"{{ query('mergelist', 'name', list1, list2) }}"
```

The results will be:

```yaml
  - name: myname01
    param01: myparam03
  - name: myname02
    param01: myparam02
    param02: myparam04
  - name: myname03
    param03: myparam03
```

### CODE REVIEW

File: tests/integration/targets/lookup_etcd3/aliases [hunk 1]
```diff
@@ -4,5 +4,6 @@ needs/file/tests/utils/constraints.txt
 needs/target/setup_etcd3
 skip/aix
 skip/osx
+skip/rhel
```
Comment (felixfontein) [inline tests/integration/targets/lookup_etcd3/aliases:7]: Please do **not** modify unrelated files.

File: changelogs/fragments/263-new_mergelist_plugin.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+features:
```
Comment (felixfontein) [inline changelogs/fragments/263-new_mergelist_plugin.yml:1]: 1. This is not a valid section.
2. New plugins and modules must not have a changelog fragment.

File: plugins/lookup/mergelist.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+# (c) 2020, Jose Angel Munoz <josea.munoz(at)gmail.com>
+# (c) 2020 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    lookup: mergelist
+    author: Jose Angel Munoz <josea.munoz@gmail.com>
+    version_added: "2.10"
```
Comment (felixfontein) [inline plugins/lookup/mergelist.py:10]: Please remove this line.

File: plugins/lookup/mergelist.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        required: True
+    notes:
+      - if key is not defined, "name" will be the default
+      - use query if you want to return a list when a single value is given.
+'''
+
+EXAMPLES = """
+  - name: Merges list1 and list2 with 'name' as key
+    debug:
+      msg: "{{ query('mergelist', list1, list2) }}"
```
Comment (felixfontein) [inline plugins/lookup/mergelist.py:26]: New plugins and modules must use FQCNs in examples.
```suggestion
      msg: "{{ query('community.general.mergelist', list1, list2) }}"
```

File: plugins/lookup/mergelist.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        - name: myname01
+          param01: myparam03
+        - name: myname02
+          param02: myparam04
+        - name: myname03
+          param03: myparam03
+
+  - name: Merges list1 and list2 with 'id' as key
+    debug:
+      msg: "{{ query('mergelist', 'id', list1, list2) }}"
```
Comment (felixfontein) [inline plugins/lookup/mergelist.py:43]: You should use keyword arguments for such parameters:
```suggestion
      msg: "{{ query('mergelist', list1, list2, key='id') }}"
```

File: tests/integration/targets/lookup_etcd3/aliases (hunk 1 above)
Comment (imjoseangel) [reply tests/integration/targets/lookup_etcd3/aliases:7]: Sorry, just testing why tests are failing with etcd.

File: changelogs/fragments/263-new_mergelist_plugin.yml (hunk 2 above)
Comment (imjoseangel) [reply changelogs/fragments/263-new_mergelist_plugin.yml]: Thanks! Removed

File: plugins/lookup/mergelist.py (hunk 3 above)
Comment (imjoseangel) [reply plugins/lookup/mergelist.py]: Thanks! Removed

File: plugins/lookup/mergelist.py (hunk 4 above)
Comment (imjoseangel) [reply plugins/lookup/mergelist.py]: Changed in all the examples

File: plugins/lookup/mergelist.py (hunk 5 above)
Comment (imjoseangel) [reply plugins/lookup/mergelist.py]: Good one. Thanks!

File: plugins/lookup/mergelist.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    lookup: mergelist
+    author: Jose Angel Munoz <josea.munoz@gmail.com>
+    short_description: merges lists from a given array key
+    description:
+      - This lookup returns the contents of two or more merged lists using a common key.
```
Comment (felixfontein) [inline plugins/lookup/mergelist.py:12]:
```suggestion
      - This lookup returns the contents of two or more merged lists of dictionaries using a common key.
```

File: plugins/lookup/mergelist.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
+# (c) 2020, Jose Angel Munoz <josea.munoz(at)gmail.com>
+# (c) 2020 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    lookup: mergelist
+    author: Jose Angel Munoz <josea.munoz@gmail.com>
+    short_description: merges lists from a given array key
```
Comment (felixfontein) [inline plugins/lookup/mergelist.py:10]:
```suggestion
    short_description: Merges lists of dictionaries by a given key
```

File: plugins/lookup/mergelist.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    description:
+      - This lookup returns the contents of two or more merged lists using a common key.
+    options:
```
Comment (felixfontein) [inline plugins/lookup/mergelist.py:13]: You need to document the option `key`.

File: plugins/lookup/mergelist.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+      _terms:
+        description: list(s) to merge
```
Comment (felixfontein) [inline plugins/lookup/mergelist.py:15]: Always use complete sentences in descriptions and notes. (`short_description` is the only exception.)
```suggestion
        description: List(s) to merge.
```

File: plugins/lookup/mergelist.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+"""
+
+from collections import defaultdict
+from ansible.plugins.lookup import LookupBase
+from ansible.errors import AnsibleError
+
+
+class LookupModule(LookupBase):
+    def run(self, terms, variables=None, **kwargs):
+
```
Comment (felixfontein) [inline plugins/lookup/mergelist.py:58]: You need to adjust the code to use the `key` keyword argument. Right now the code will not work with the second example in the documentation.

File: plugins/lookup/mergelist.py (hunk 6 above)
Comment (imjoseangel) [reply plugins/lookup/mergelist.py]: Thank you! Changed

File: plugins/lookup/mergelist.py (hunk 7 above)
Comment (imjoseangel) [reply plugins/lookup/mergelist.py]: Changed

File: plugins/lookup/mergelist.py (hunk 8 above)
Comment (imjoseangel) [reply plugins/lookup/mergelist.py:13]: Added. Silly me, I didn't realize carefully about your comment. Thank you!

File: plugins/lookup/mergelist.py (hunk 10 above)
Comment (imjoseangel) [reply plugins/lookup/mergelist.py:94]: Thank you!!! I din't realize about that change. Now with key and with better integration tests

**Comment by imjoseangel** [comment]
> If you want to improve chances of this getting merged, you should explain **why** such a lookup is needed in Ansible. For which use-cases is this needed, that cannot be solved by other means (using existing lookups and/or filters)?

Thanks @felixfontein . I have updated the aditional information with:

When creating complex and long structures there are ways to merge or replace one to one dictionaries starting from their head. There is no a simple way to merge small fragments of different dictionaries from a given master key.

For instance: A keyvault ACL in Azure can be unreadable. To keep a consistency and merge new configurations and don't break the existing ones, something like this can be performed:

```yaml
q('mergelist',
       _keyvaultinfo.keyvaults[0].access_policies | default([]),
        keyvault_objectid.keyvaultsp | default([]) ) |
          json_query('[*].{
            tenant_id: tenant_id, 
            object_id: object_id,
            certificates: permissions.certificates,
            keys: permissions.keys,
            secrets: permissions.secrets
            }', 
           key='object_id')
```
Something like this without this plugin is not easily achievable.

**Comment by imjoseangel** [comment]
@felixfontein Question,

It is quite frustrating that Shippable tests fail always with lookup plugin test_lookup_etcd3 but nothing to do with this new plugin. Could you please advice what is the best alias and/or process to go?

Thanks

**Comment by imjoseangel** [comment]
Hi @felixfontein

Any news about merging this feature?

What is the next steps and procedures for this Ansible collection?

Thank you very much!

**Comment by felixfontein** [comment]
@imjoseangel two things:
1. we only merge new plugins/modules if they look useful to a larger number of people. Right now, I'm not sure whether that's the case, but if other people would request such a plugin as well, I won't block it. Maybe ask in the support mailing list (ansible-project@googlegroups.com) if others would be interested in such a lookup as well. Also, maybe other people active in this repository would be interested, but right now I didn't see anyone else commenting here (or in the issue).
2. all new plugins and modules must have tests, preferably integration tests. Look for example at https://github.com/ansible-collections/community.general/tree/master/tests/integration/targets/lookup_cartesian for how tests for such a plugin can be added.

Also, what this plugin does can essentially be done with pure Jinja2/Ansible: `q('mergelist', list_of_dicts_1, list_of_dicts_2, list_of_dicts_3, key='key')` can be essentially replaced by
```.j2
(list_of_dicts_1 | map(attribute='key') | zip(list_of_dicts_1)) | items2dict
| combine(
    (list_of_dicts_2 | map(attribute='key') | zip(list_of_dicts_2)) | items2dict,
    recursive=true)
| combine(
    (list_of_dicts_3 | map(attribute='key') | zip(list_of_dicts_3)) | items2dict,
    recursive=true)
.values() | list
```
(The behavior of merging dict entries inside the dicts is different from what your plugin does.)

If replacing such structures would be a common problem, I would agree that this should be merged. But I'm not sure that they are very common.

**Comment by imjoseangel** [comment]
Thank you very much @felixfontein . Let's wait then.

Regarding point 2, it is already done. Find [here](https://github.com/ansible-collections/community.general/pull/263/files#diff-aef8e4a6c39f49e29bc7455cd801bb9f)

Thank you again!

**Comment by Krakoslabs** [comment]
👍

**Comment by lupocito** [comment]
This could be great! 👍

**Comment by gundalow** [comment]
@imjoseangel Thanks for adding integration test

@Krakoslabs @lupocito Hi, are you able to download this patch and test it locally? Getting some feedback on if this would would allow us to merge. Thanks in advance

**Comment by felixfontein** [comment]
@Krakoslabs @lupocito please also state what you would use this plugin for. I'd really like to see some more applications.

**Comment by vbotka** [comment]
FWIW, standard filters are pretty awkward here. For example

```yaml
    - set_fact:
        my_dict1: "{{ my_dict1|default({})|
                      combine({item.0: item.1}) }}"
      loop: "{{ (list1 + list2)|groupby('name') }}"
    - debug:
        var: my_dict1

    - set_fact:
        my_dict2: "{{ my_dict2|default({})|
                      combine({item.0.key: item.1}, recursive=True) }}"
      with_subelements:
        - "{{ my_dict1|dict2items }}"
        - value
    - debug:
        var: my_dict2
```
give

```yaml
  my_dict1:
    myname01:
    - name: myname01
      param01: myparam01
    - name: myname01
      param01: myparam03
    myname02:
    - name: myname02
      param01: myparam02
    - name: myname02
      param02: myparam04
    myname03:
    - name: myname03
      param03: myparam03

  my_dict2:
    myname01:
      name: myname01
      param01: myparam03
    myname02:
      name: myname02
      param01: myparam02
      param02: myparam04
    myname03:
      name: myname03
      param03: myparam03
```

**Comment by vbotka** [comment]
FWIW, a simple filter does the job as well. For example

```python
shell> cat filter_plugins/list_filters.py
from collections import defaultdict

def lists_merge(l1, l2, index):
    d = defaultdict(dict)
    for l in (l1, l2):
        for elem in l:
            d[elem[index]].update(elem)
            l3 = d.values()
    return l3
                                                        
class FilterModule(object):
    ''' Ansible filters. Interface to Python list methods.'''

    def filters(self):
        return {
            'lists_merge': lists_merge
        }
```

gives the required result

```yaml
    - debug:
        msg: "{{ list1|lists_merge(list2, 'name')|list }}"
```

```yaml
  msg:
  - name: myname01
    param01: myparam03
  - name: myname02
    param01: myparam02
    param02: myparam04
  - name: myname03
    param03: myparam03
```

**Comment by imjoseangel** [comment]
Thanks @vbotka for your feedback. That is exactly the idea.

**Comment by felixfontein** [comment]
@imjoseangel @vbotka so now we have two PRs for almost the same functionality, once as a lookup (this PR) and once as a filter (#604).

The lookup has documentation (filter plugins currently do not have proper docs), allows multiple lists, and has more tests. The filter has a nicer syntax (feels more like Jinja2 than a lookup). So how do you want to proceed?

**Comment by imjoseangel** [comment]
Honestly I like both ideas, I let you choose guys 😁

**Comment by felixfontein** [comment]
@vbotka @imjoseangel it would be nice if you (and other persons who want to have this functionality) could talk about this and decide how to proceed. I'm happy to help reviewing and merge (once everything's fine) any of these PRs, but I'm not going to make a decision for you which of the PRs should be continued.

**Comment by imjoseangel** [comment]
Thanks @felixfontein ,

@vbotka The initial requirement was having a lookup plugin. Personally I prefer that approach as it is closer to dicts and list behavior. We can open an IRC channel together with @Krakoslabs  and @lupocito to discuss the best option. 

In the case of implementing the filter plugin we need to work in some unit tests for the filter part.

@vbotka Let us know your thoughts

Regards

**Comment by felixfontein** [comment]
Where does this requirement of having a lookup plugin coming from?

**Comment by imjoseangel** [comment]
> Where does this requirement of having a lookup plugin coming from?

Nah, It was just a personal thought and requirement coming from the **Team** from I was part of. I opened the issue that way based in our experience and how we managed configuration. Using _lookup_ and _query_ options gave us more flexibility (Dicts and List outcome) and was easier to understand than using a filters. Mainly for sake of readability and functionality.

I thought that having that requirement will help more users than the filter. From the new comments and ideas, I'm not sure what functionality can give more value to the users although I still more in the side of the lookup/query.

To me, a filter is cool to convert for instance a sastoken or other quick conversions but not for the scenario I was describing in the issue. This is also is a personal thought.

Saying this, if we go for the filter option, I want to be sure that it will help as much users as possible. To do this, we need to enforce some unit testing and good documentation and I will help on that if that is the case.

**Comment by imjoseangel** [comment]
@vbotka @Krakoslabs @lupocito Any thoughts?

Thank you!

**Comment by felixfontein** [comment]
It would be good to settle this soon. 1.1.0 will be released on 2020-08-18, and if this (or the other PR) should go in, it needs to be merged by then.

Pinging @vbotka @Krakoslabs @lupocito again

**Comment by imjoseangel** [comment]
Kindly reminder. @felixfontein How do you manage these open issues. Because there is not enough support, do you think it is good to close it?

Thank you!

**Comment by imjoseangel** [comment]
Good Example to use mergelist I have just implemented:

I needed to create an adhoc inventory like this:

```yaml
---
- name: Add linux servers to inventory
  add_host:
    name: "{{ linuxinventory.name }}"
    ansible_host: "{{ linuxinventory.hostname }}"
    ansible_user: "{{ linuxinventory.username }}"
    ansible_python_interpreter: "/usr/bin/env python"
    groups: "{{ linuxinventory.groups | default('linux') }}"
    instanceName: "{{ linuxinventory.instanceName | default('') }}"
    dbPass: "{{ linuxinventory.dbpass | default('') }}"
  loop: "{{ toolbox.inventory | flatten(levels=1) }}"
  loop_control:
    loop_var: linuxinventory
```

>**Note:** no_log is missing in some of the examples for the sake of simplicity

Mainly the vars are comming from two places:

```yaml
common:
  myservers:
    inventory:
      - name: MYSERVER01
        hostname: MYSERVER01.example.com
        username: myuser
        groups:
          - group1
          - group2
          - group3
        instanceName: MYINSTANCE
```

and

```yaml
---
secrets:
  myservers:
    inventory:
      - name: MYSERVER01
        dbpass:
          dbschemapass: myencryptedpass1
          dbstmpchemapass: myencryptedpass1
```

The second file is encrypted with `ansible-vault`

Well, the playbook reads both variable files (One standard and one secret) and pass it to the role like:

```yaml
    - name: Create inventory
      include_role:
        name: toolbox
      vars:
        toolbox:
          inventory: "{{ query('mergelist', common.myservers.inventory, secrets.myservers.inventory, key='name') | default('') }}"
        action_toolbox: buildinventory
      no_log: true
```

Creating the inventory with the needed vars for deployment.

**Comment by Andersson007** [comment]
@felixfontein PTAL

**Comment by gundalow** [comment]
@imjoseangel @Krakoslabs @lupocito @vbotka What's your view, use filter, or lookup

https://github.com/ansible-collections/community.general/pull/263
https://github.com/ansible-collections/community.general/pull/604

**Comment by dmsimard** [comment]
> @imjoseangel @Krakoslabs @lupocito @vbotka What's your view, use filter, or lookup
> 
> #263
> #604

From a functional point of view, a lookup generally queries something (a system, an application, an API, etc.) and then returns data while a filter takes existing input data and munges it into something else.

With that in mind a filter would be a better fit for this.

**Comment by imjoseangel** [comment]
> With that in mind a filter would be a better fit for this.

Fine to me. Go ahead, close #263 and merge #604 . I think this will be a pretty cool functionality for 2.11

**Comment by felixfontein** [comment]
Closing in favor of #604.

Thanks for your work on this, @imjoseangel!
